src/lib/log.c

```diff
@@ -0,0 +1,50 @@
+#include <asl.h>
+#include <pwd.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <time.h>
+
+#include <SystemConfiguration/SystemConfiguration.h>
+
+#include <xhyve/log.h>
+
+static aslclient log_client = NULL;
+static aslmsg log_msg = NULL;
+
+static unsigned char buf[4096];
+/* Index of the _next_ character to insert in the buffer. */
+static size_t buf_idx = 0;
+
+/* asl is deprecated in favor of os_log starting with macOS 10.12.  */
+#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
+
+/* Initialize ASL logger and local buffer. */
+void log_init(void)
+{
+	log_client = asl_open(NULL, NULL, 0);
+	log_msg = asl_new(ASL_TYPE_MSG);
+}
+
+
+/* Send the content of the buffer to the logger. */
+static void log_flush(void)
+{
+	buf[buf_idx] = 0;
+	asl_log(log_client, log_msg, ASL_LEVEL_NOTICE, "%s", buf);
+	buf_idx = 0;
```

Comment:
Do we care that `\n\n\n\n` will log a bunch of blank messages? A simple `if (buf_idx == 0) return` check at the top would fix that if we did.

Reply:
Actually, that's what I wanted, to keep the current output.

<img width="1570" alt="image" src="https://user-images.githubusercontent.com/232441/38498401-9a686f4a-3c04-11e8-96ea-b70406e1a600.png">

Reply:
Ah, yes, that's a good point.

Reply:
BTW, would you happen to know why we have two whales?  Maybe it's D4D specific, I have not checked yet, but I'm kinda worried by the backslashes having been processed.

<img width="1244" alt="image" src="https://user-images.githubusercontent.com/232441/38498675-76bf6494-3c05-11e8-9c81-2f81894f4d06.png">

Reply:
There was some strangeness wrt how/where motd was handled in linuxkit but that was more than a year ago, I don't recall having seen anything like this recently.

Are you running the getty in a container or within the root? Or even both (which might explain things)?

I presume the console is functional (i.e. you can login and type normally)? If it isn't that might suggest you've got two `getty` processes somehow.

Only other thought I have is if you had configured two serial ports you would get a getty and motd on both of them, but the patches here would combine them into the logs without indicating which was which. If that were the case I'd expect them to be more mixed up though.

Reply:
Before these patches, both whales were already in `console-ring`.  I have not dug the issue yet, so I can't even answer to your questions now.  It was just in case it rang a bell.  Thanks for the hints, I'll use them when I focus on the twin-whale-issue.

Reply:
Actually looks like I misremembered. In LinuxKit (modern ones at least) it is actually an initscript which cats `/etc/issue` (not `motd`) to any device with a `console=` on the kernel command line.

Looks like `agetty` (which LinuxKit uses) will also print `/etc/issue` but if it is in a container then that is a different `/etc/issue`, unless you bind it in or arrange to use a different getty container. LinuxKit demos don't have that bind but maybe you do? Or if you were running getty on the host you would see it.